modules/substrate/src/justification.rs

```diff
@@ -0,0 +1,329 @@
+
+/// Justification verification error.
+#[derive(RuntimeDebug, PartialEq)]
+pub enum Error {
+	/// Failed to decode justification.
+	JustificationDecode,
+	/// Justification is finalizing unexpected header.
+	InvalidJustificationTarget,
+	/// Invalid commit in justification.
+	InvalidJustificationCommit,
+	/// Justification has invalid authority singature.
+	InvalidAuthoritySignature,
+	/// The justification has precommit for the header that has no route from the target header.
+	InvalidPrecommitAncestryProof,
+	/// The justification has 'unused' headers in its precommit ancestries.
+	InvalidPrecommitAncestries,
+}
+
+/// Verify that justification, that is generated by given authority set, finalizes given header.
+pub fn verify_justification<Header: HeaderT>(
+	finalized_target: (Header::Hash, Header::Number),
+	authorities_set_id: SetId,
+	authorities_set: VoterSet<AuthorityId>,
+	raw_justification: &[u8],
+) -> Result<(), Error>
+where
+	Header::Number: finality_grandpa::BlockNumberOps,
+{
+	// decode justification first
+	let justification =
+		GrandpaJustification::<Header>::decode(&mut &raw_justification[..]).map_err(|_| Error::JustificationDecode)?;
+
+	// ensure that it is justification for the expected header
+	if (justification.commit.target_hash, justification.commit.target_number) != finalized_target {
+		return Err(Error::InvalidJustificationTarget);
+	}
+
+	// validate commit of the justification (it just assumes all signatures are valid)
+	let ancestry_chain = AncestryChain::new(&justification.votes_ancestries);
+	match finality_grandpa::validate_commit(&justification.commit, &authorities_set, &ancestry_chain) {
+		Ok(ref result) if result.ghost().is_some() => {}
+		_ => return Err(Error::InvalidJustificationCommit),
+	}
+
+	// now that we know that the commit is correct, check authorities signatures
+	let mut buf = Vec::new();
+	let mut visited_hashes = BTreeSet::new();
+	for signed in &justification.commit.precommits {
+		if !sp_finality_grandpa::check_message_signature_with_buffer(
+			&finality_grandpa::Message::Precommit(signed.precommit.clone()),
+			&signed.id,
+			&signed.signature,
+			justification.round,
+			authorities_set_id,
+			&mut buf,
+		) {
+			return Err(Error::InvalidAuthoritySignature);
+		}
+
+		if justification.commit.target_hash == signed.precommit.target_hash {
+			continue;
+		}
+
+		match ancestry_chain.ancestry(justification.commit.target_hash, signed.precommit.target_hash) {
+			Ok(route) => {
+				// ancestry starts from parent hash but the precommit target hash has been visited
+				visited_hashes.insert(signed.precommit.target_hash);
+				visited_hashes.extend(route);
+			}
+			_ => {
+				// could this happen in practice? I don't think so, but original code has this check
+				return Err(Error::InvalidPrecommitAncestryProof);
+			}
+		}
+	}
+
+	let ancestry_hashes = justification
+		.votes_ancestries
+		.iter()
+		.map(|h: &Header| h.hash())
+		.collect();
```

Comment:
To avoid collecting to a `Vec` we can just `.zip` with `visited_hashes` and then verify in a loop.

Reply:
Using `zip()`assumes that both iterators are sorted in the same order. I don't know if it is true.

Reply:
Ah, fair enough. I thought we compare `Vec`s.